Allow use of user-specified private galaxy server with fallback

AlanCoding opened the pull request:
##### SUMMARY
Allow users to use a custom URL / authentication to a Galaxy server.

This will still set up the public galaxy server as a fallback.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
6.1.0
```


##### ADDITIONAL INFORMATION
Several things have been completed in Ansible core to open the door for this.

https://github.com/ansible/ansible/issues/61247

AlanCoding committed:
Initial env var implementation of private galaxy server

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -318,6 +318,15 @@ def _load_default_license_from_file():
 )
 
+register(
+    'PROJECT_UPDATE_VVV',
```
🙏 you are a saint, @AlanCoding

wenottingham replied:
Will this leak details on the creds that it shouldn't?

AlanCoding replied:
Any project update stdout still goes through the URL cleaner. I'm more worked about the cleaner bugging out than I am worried about leaking secrets.

Having this setting, however, would allow us to expand test coverage for URL sanitization. I will make a note of that.

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -338,6 +347,52 @@ def _load_default_license_from_file():
+    field_class=fields.CharField,
+    label=_('Private Galaxy Server Token'),
+    help_text=_('For using a private galaxy server at higher precedence than the public Ansible Galaxy. '
+                'The username to use for basic authentication against the Galaxy instance, '
```
This docstring needs some love

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1830,6 +1830,19 @@ def build_env(self, project_update, private_data_dir, isolated=False, private_da
+        if private_galaxy_url:
+            # set up the fallback server, which is the normal Ansible Galaxy
+            env['ANSIBLE_GALAXY_SERVER_GALAXY_URL'] = 'https://galaxy.ansible.com'
+            env['ANSIBLE_GALAXY_SERVER_PRIVATE_GALAXY_URL'] = private_galaxy_url
```
Do you need this line given `url` on L1838 in the loop?

AlanCoding replied:
huh... well... I guess not

ryanpetrello commented on awx/ui/client/src/configuration/forms/jobs-form/configuration-jobs.form.js:
```diff
@@ -58,12 +58,19 @@ export default ['i18n', function(i18n) {
             AWX_COLLECTIONS_ENABLED: {
                 type: 'toggleSwitch',
             },
+            PRIVATE_GALAXY_URL: {
```
Probably needs the other fields, too?

AlanCoding committed:
Avoid redacting Galaxy URLs

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -338,6 +347,52 @@ def _load_default_license_from_file():
+    'PRIVATE_GALAXY_PASSWORD',
+    field_class=fields.CharField,
+    label=_('Private Galaxy Server Password'),
+    help_text=_('For using a private galaxy server at higher precedence than the public Ansible Galaxy. '
```
`encrypted=True`

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -338,6 +347,52 @@ def _load_default_license_from_file():
+    help_text=_('For using a private galaxy server at higher precedence than the public Ansible Galaxy. '
+                'The username to use for basic authentication against the Galaxy instance, '
+                'this is mutually exclusive with corresponding username and password settings.'),
+    category=_('Jobs'),
```
`encrypted=True`

AlanCoding committed:
Finish adding settings to UI

AlanCoding commented:
Still TODO:

validation of settings is needed.

If the URL is blank, then none of the other fields should be accepted.

If the token is provided, username/pass should not be accepted, and vice versa.

wenottingham commented on awx/main/conf.py:
```diff
@@ -338,6 +347,59 @@ def _load_default_license_from_file():
+    field_class=fields.URLField,
+    required=False,
+    allow_blank=True,
+    label=_('Private Galaxy Server URL'),
```
Probably can't call it that.

AlanCoding replied:
Is the problem word "Private" or "Galaxy"? This is used to configure the Ansible core settings related to galaxy servers, so it doesn't seem possible to avoid the latter.

https://github.com/ansible/ansible/blob/devel/lib/ansible/config/base.yml#L1347

AlanCoding committed:
validate galaxy server settings

involves some changes to the redact code

AlanCoding commented on awx/main/conf.py:
```diff
@@ -608,4 +670,51 @@ def logging_validate(serializer, attrs):
+        galaxy_data[subfield] = _new_value('PRIVATE_GALAXY_{}'.format(subfield.upper()))
+    errors = {}
+    print('galaxy data')
+    print(galaxy_data)
```
remove

AlanCoding committed:
Rename private to primary in galaxy settings

use a setting for the public galaxy URL

chrismeyersfsu commented:
I'd check with Dylan and rich to see if we want an open ended private galaxy url vs. hard coded automation hub. Also point it out on demo day.

AlanCoding committed:
Describe usage of primary galaxy server in docs

AlanCoding commented on docs/collections.md:
```diff
@@ -53,3 +53,31 @@ Example of tmp directory where job is running:
+how Galaxy servers are configured in Ansible in general see:
+
+https://docs.ansible.com/ansible/devel/dev_guide/collections_tech_preview.html
+(if "devel" link goes stale in the future, it is for Ansible 2.9)
```
🚚 https://docs.ansible.com/ansible/devel/user_guide/collections_using.html

AlanCoding commented:
recheck

AlanCoding committed:
Redact env vars for Galaxy token or password

ryanpetrello reviewed (approved):
This makes sense to me.  Once we've verified it against a private Galaxy install, I say :shipit:

AlanCoding committed:
Add setting for auth_url

Also adjust public galaxy URL setting to
allow using only the primary Galaxy server

Include auth_url in token exclusivity validation

AlanCoding committed:
update docs for galaxy auth URL material

AlanCoding committed:
Add more galaxy server param validation

AlanCoding committed:
bump required version to 2.10

AlanCoding committed:
Revert "bump required version to 2.10"

This reverts commit e0e9c8321b223b8cc4b6aceb7225a2bab9ba109c.

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -700,4 +779,75 @@ def logging_validate(serializer, attrs):
+            if (not serializer.instance) or (not getattr(serializer.instance, setting_name, '')):
+                # new auth is applied, so check if compatible with version
+                from awx.main.utils import get_ansible_version
+                current_version = get_ansible_version()
```
Are there any performance implications to calling this constantly in settings validation?  How is this value fetch cached?

ryanpetrello replied:
Looks like it caches it in-memory every 60 seconds, so that should be good enough for "once per-request".  Does this validation only get called when the setting value is _accessed_, though?  I want to make sure we're not introducing this call once per _every_ request that touches `django.conf.settings` in some way.

AlanCoding replied:
yep, the answer here is that we only call this if 2 conditions are met - that the new field value is non-blank, and the old field value is blank. I was worried about this exact thing. Provided I've written this correctly, I think it should very specifically avoid anything like this

> I want to make sure we're not introducing this call once per every request that touches django.conf.settings in some way.
